File: Wireless Flash/src/models/shared_models/FileTransfer.java
package models.shared_models;

import java.io.*;


/**
 * This class is used to transfer and receive files
 */
public class FileTransfer {
	private long transferredFileSize = 0;
	private File firstFile;
	private boolean work=true;
	
	private static final int BUFFER = 1024 * 1024 * 4;

	/**
	 * method used to recursively upload files/folders
	 * @param outputStreamStrings is output stream for strings
	 * @param outputStreamBytes is output stream for bytes
	 * @param file is the main file/folder to be uploaded 
	 * @param mainPath is the parents path (to establish relationship of files)
	 */
	public void sendFiles(DataOutputStream outputStreamStrings,
			DataOutputStream outputStreamBytes,
			File file, String mainPath) {
		
		try {
			//check if pipe was broken
			if(!work)
				return;

			BasicFileData basicFileData = new BasicFileData(file);

			basicFileData.setPath(basicFileData.getPath().substring(mainPath.length()));
			
			
			if(basicFileData.getPath().startsWith("\\") || basicFileData.getPath().startsWith("/"))
				basicFileData.setPath(basicFileData.getPath().substring(1));

			String jsonFile = JsonParser.basicFileDataToJson(basicFileData);
			
			outputStreamStrings.write(jsonFile.getBytes("UTF-8"));

			outputStreamStrings.writeBytes("\n");
			
			outputStreamStrings.flush();
			
			if(file.isDirectory()) {
				File[] list = file.listFiles();
				for(int i = 0; i < list.length; i++)
					sendFiles(outputStreamStrings, outputStreamBytes, list[i], mainPath);
			}else{
				FileInputStream fileData = new FileInputStream(file);
				byte[] buffer = new byte[BUFFER];
				long size = file.length();
				while(size != 0) {
					if(size >= buffer.length) {
						fileData.read(buffer, 0, buffer.length);
						outputStreamBytes.write(buffer, 0, buffer.length);
						size -= buffer.length;
						this.transferredFileSize += buffer.length;
					}
					else {
						fileData.read(buffer, 0, (int)size);
						outputStreamBytes.write(buffer, 0, (int)size);
						this.transferredFileSize += size;
						size = 0;
					}
					outputStreamBytes.flush();
				}
				fileData.close();
			}
		}catch(Exception e) {
			work = false; //pipe was broken
			e.printStackTrace();
		}
	}

	/**
	 * method used to download files/folders
	 * @param byteStream is input stream to receive bytes from
	 * @param inputStream is input stream to receive strings from
	 * @param path is location to save data under
	 */
	public void receiveFiles(DataInputStream byteStream, BufferedReader inputStream, String path){
		FileOutputStream output = null;
		try {
			for(String temp; (temp = inputStream.readLine()) != null; ) {
				BasicFileData[] tempBasicFileData = JsonParser.JsonToBasicFileData(temp);
				BasicFileData basicFileData = null;

				if(tempBasicFileData != null)
					basicFileData = tempBasicFileData[0];
				else
					throw new Exception("No Meta Data");

				if(basicFileData.isDirectory()) {
					File file = new File(path+basicFileData.getPath());
					
					//check if file is the first to be received
					if(firstFile == null)
						firstFile = file;
					
					file.mkdirs();
				}
				else {

					output = new FileOutputStream(path+basicFileData.getPath());
					long size = basicFileData.getSize();
					byte[] buffer = new byte[BUFFER];

					//check if file is the first to be received
					if(firstFile == null)
						firstFile = new File(path+basicFileData.getPath());
					
					while(size > 0) {
						int bytesRead = 0;

						if(size > BUFFER)
							bytesRead = byteStream.read(buffer, 0, BUFFER);
						else
							bytesRead = byteStream.read(buffer, 0, (int)size);

						if (bytesRead != -1) {
							this.transferredFileSize += bytesRead;
							size -= bytesRead;
							output.write(buffer, 0, bytesRead);
						}else
							break;

					}
					output.close();
				}
			}
		}catch(Exception e) {	
			try {output.close();} catch (IOException e1) {}
			deleteFile(firstFile);
			e.printStackTrace();
		}
	}
	
	/**
	 * This method is used to calculate Total size for files/folders
	 * @param file is a file/folder
	 * @return the total size of the file/folder
	 */
	public long calculateSize(File file) {
		long sum = 0;
		if(file.isDirectory()) {
			for(File temp : file.listFiles()) {
				if(temp.isDirectory())
					sum += calculateSize(temp);
				else
					sum += temp.length();
			}
		}
		else{
			sum += file.length();
		}
		return sum;
	}
	
	/**
	 * get method for transferredFileSize
	 * @return the number of bytes that have been written/read so far
	 */
	public long getTransferredFileSize() {
		return transferredFileSize;
	}


	/**
	 * method used to delete a file/folder
	 * @param file is file/folder to be deleted
	 */
	private static void deleteFile(File file) {
		if(file == null || !file.exists())
			return;
		try {
			if(file.isDirectory()) {
				for(File sub : file.listFiles()) {
					if(sub.isDirectory())
						deleteFile(sub);
					sub.delete();
				}
			}
			file.delete();	
		}catch(Exception e){
			e.printStackTrace();
		}
	}
	
}

File: Wireless Flash/src/models/shared_models/FileTransfer.kt
package models.shared_models


import java.io.*


/**
 * This class is used to transfer and receive files
 */
class FileTransfer {
    /**
     * get method for transferredFileSize
     * @return the number of bytes that have been written/read so far
     */
    var transferredFileSize: Long = 0
        private set
    private var firstFile: File? = null
    private var work = true

    /**
     * method used to recursively upload files/folders
     * @param outputStreamStrings is output stream for strings
     * @param outputStreamBytes is output stream for bytes
     * @param file is the main file/folder to be uploaded
     * @param mainPath is the parents path (to establish relationship of files)
     */
    fun sendFiles(outputStreamStrings: DataOutputStream,
                  outputStreamBytes: DataOutputStream,
                  file: File, mainPath: String) {

        try {
            //check if pipe was broken
            if (!work)
                return

            val basicFileData = BasicFileData(file)

            basicFileData.path = basicFileData.path!!.substring(mainPath.length)


            if (basicFileData.path!!.startsWith("\\") || basicFileData.path!!.startsWith("/"))
                basicFileData.path = basicFileData.path!!.substring(1)

            val jsonFile = JsonParser.basicFileDataToJson(Array<BasicFileData?>(1) {basicFileData })

            outputStreamStrings.write(jsonFile.toByteArray(charset("UTF-8")))

            outputStreamStrings.writeBytes("\n")

            outputStreamStrings.flush()

            if (file.isDirectory) {
                val list = file.listFiles()
                for (i in list!!.indices)
                    sendFiles(outputStreamStrings, outputStreamBytes, list[i], mainPath)
            } else {
                val fileData = FileInputStream(file)
                val buffer = ByteArray(BUFFER)
                var size = file.length()
                while (size != 0L) {
                    if (size >= buffer.size) {
                        fileData.read(buffer, 0, buffer.size)
                        outputStreamBytes.write(buffer, 0, buffer.size)
                        size -= buffer.size.toLong()
                        this.transferredFileSize += buffer.size.toLong()
                    } else {
                        fileData.read(buffer, 0, size.toInt())
                        outputStreamBytes.write(buffer, 0, size.toInt())
                        this.transferredFileSize += size
                        size = 0
                    }
                    outputStreamBytes.flush()
                }
                fileData.close()
            }
        } catch (e: Exception) {
            work = false //pipe was broken
            e.printStackTrace()
        }

    }

    /**
     * method used to download files/folders
     * @param byteStream is input stream to receive bytes from
     * @param inputStream is input stream to receive strings from
     * @param path is location to save data under
     */
    fun receiveFiles(byteStream: DataInputStream, inputStream: BufferedReader, path: String) {
        var output: FileOutputStream? = null
        try {
            var temp: String? =inputStream.readLine()
            while (temp != null) {
                val tempBasicFileData = JsonParser.JsonToBasicFileData(temp)
                var basicFileData: BasicFileData?

                if (tempBasicFileData != null)
                    basicFileData = tempBasicFileData[0]
                else
                    throw Exception("No Meta Data")

                if (basicFileData!!.isDirectory) {
                    val file = File(path + basicFileData.path!!)

                    //check if file is the first to be received
                    if (firstFile == null)
                        firstFile = file

                    file.mkdirs()
                } else {

                    output = FileOutputStream(path + basicFileData.path!!)
                    var size = basicFileData.size
                    val buffer = ByteArray(BUFFER)

                    //check if file is the first to be received
                    if (firstFile == null)
                        firstFile = File(path + basicFileData.path!!)

                    while (size > 0) {
                        val bytesRead: Int = if (size > BUFFER)
                                                byteStream.read(buffer, 0, BUFFER)
                                             else
                                                byteStream.read(buffer, 0, size.toInt())

                        if (bytesRead != -1) {
                            this.transferredFileSize += bytesRead.toLong()
                            size -= bytesRead.toLong()
                            output.write(buffer, 0, bytesRead)
                        } else
                            break

                    }
                    output.close()
                }
                temp = inputStream.readLine()
            }
        } catch (e: Exception) {
            try {
                output!!.close()
            } catch (e1: IOException) {
            }

            deleteFile(firstFile)
            e.printStackTrace()
        }

    }

    /**
     * This method is used to calculate Total size for files/folders
     * @param file is a file/folder
     * @return the total size of the file/folder
     */
    fun calculateSize(file: File): Long {
        var sum: Long = 0
        if (file.isDirectory) {
            for (temp in file.listFiles()!!) {
                if (temp.isDirectory)
                    sum += calculateSize(temp)
                else
                    sum += temp.length()
            }
        } else {
            sum += file.length()
        }
        return sum
    }

    companion object {

        private val BUFFER = 1024 * 1024 * 4


        /**
         * method used to delete a file/folder
         * @param file is file/folder to be deleted
         */
        private fun deleteFile(file: File?) {
            if (file == null || !file.exists())
                return
            try {
                if (file.isDirectory) {
                    for (sub in file.listFiles()!!) {
                        if (sub.isDirectory)
                            deleteFile(sub)
                        sub.delete()
                    }
                }
                file.delete()
            } catch (e: Exception) {
                e.printStackTrace()
            }

        }
    }

}
